rpc/handlers.go

```diff
@@ -775,9 +776,70 @@ func getAddInvokeTxCode(err error) int {
 		return jsonrpc.InvalidParams
 	case strings.Contains(err.Error(), "not supported. Supported versions"):
 		return jsonrpc.InvalidParams
-	case strings.Contains(err.Error(), "max_fee must be bigger than 0.\n0 >= "):
+	case strings.Contains(err.Error(), "max_fee must be bigger than 0"):
 		return jsonrpc.InvalidParams
 	default:
 		return jsonrpc.InternalError
 	}
 }
+
+// AddDeclareTransaction relays a declare transaction to the gateway.
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_write_api.json#L39
+// Note: The gateway expects the sierra_program to be gzip compressed and 64-base encoded. We perform this operation,
+// and then relay the transaction to the gateway.
+func (h *Handler) AddDeclareTransaction(declareTx *json.RawMessage) (*DeclareTxResponse, *jsonrpc.Error) {
+	var request map[string]interface{}
+	err := json.Unmarshal([]byte(*declareTx), &request)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+
+	if version, ok := request["version"]; ok && version == "0x2" {
+		contractClass, ok := request["contract_class"].(map[string]interface{})
+		if !ok {
+			return nil, jsonrpc.Err(jsonrpc.InvalidParams, errors.New("{'contract_class': ['Missing data for required field.']}").Error())
+		}
+		sierraProg, ok := contractClass["sierra_program"]
+		if !ok {
+			return nil, jsonrpc.Err(jsonrpc.InvalidParams, errors.New("{'sierra_program': ['Missing data for required field.']}").Error())
```

Comment:
I think we can pass in the string directly. No need for `errors.New(...).Error()`. Same for the error a few lines up.

Reply:
Also: is there a reason why the strings are formatted like json objects instead of normal English sentence? It may be the right thing to do, I'm just wondering if there's a specific reason :)

Reply:
> Also: is there a reason why the strings are formatted like json objects instead of normal English sentence? It may be the right thing to do, I'm just wondering if there's a specific reason :)

Great question, I use this json-format just because it's what the sequencer would return in this scenario.